### PR TITLE
Enable deployment via GH actions

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR adds in Github Actions workflows to deploy packages to TestPyPI on release branches and to PyPI itself whenever a tag is pushed.

## Motivation and Context

We need these in place for releases. When switching from CircleCI to Github Actions we chose not to port the deployment step, opting to instead wait until we had an actual release since we knew we would need to work through some issues. This PR handles those changes.

## Checklist:

- [x] I am familiar with the [Contributing Guidelines](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [Contributor Agreement](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [ ] The [package documentation](https://github.com/glotzerlab/signac/tree/master/doc) and [framework documentation](https://docs.signac.io/) in [signac-docs](https://github.com/glotzerlab/signac-docs) are up to date with these changes.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added any related issue and pull request numbers for future reference.

### CODE REVIEW

**Comment by vyasr** [comment]
If reviewers are interested you can look into the actions tab to see the last few runs of both the TestPyPI and normal PyPI upload workflows. Up until the last few commits I had it triggering on every push to the branch and disabled the final twine uploads, so you can see the rest of the output there. You can also see that in the two cases where I did enable TestPyPI uploads [we did get new packages](https://test.pypi.org/project/signac/1.7.0.dev1/#history) (dev0 and dev1 releases).

File: .github/workflows/deploy-pypi.yml
Comment (bdice) [inline .github/workflows/deploy-pypi.yml:6]: Can we match on something closer to SemVer number schemes? Like regex `v[0-9]+\.[0-9]+\.[0-9]+`. Not sure if there is a similar glob.

File: .github/workflows/deploy-test-pypi.yml
Comment (bdice) [inline .github/workflows/deploy-test-pypi.yml:6]: I think this needs a slash? Or a dash?
```suggestion
      - 'release/*'
```

File: .github/workflows/deploy-pypi.yml
Comment (bdice) [inline .github/workflows/deploy-pypi.yml:14]: Reminder to fix this before merging.
```suggestion
    uses: glotzerlab/signac/.github/workflows/publish-packages.yml
```

File: .github/workflows/deploy-test-pypi.yml
Comment (bdice) [inline .github/workflows/deploy-test-pypi.yml:14]:
```suggestion
    uses: glotzerlab/signac/.github/workflows/publish-packages.yml
```

File: .github/workflows/publish-packages.yml
Comment (bdice) [inline .github/workflows/publish-packages.yml:30]: Let’s remove the second half of this comment that suggests an alternative. I like the strategy as written.

File: .github/workflows/publish-packages.yml
Comment (bdice) [inline .github/workflows/publish-packages.yml:38]: Why does this go to `--user` but the other requirements do not?

File: .github/workflows/publish-packages.yml
Comment (bdice) [inline .github/workflows/publish-packages.yml:44]: Delete. I prefer manual installation of requirements for this.

**Comment by bdice** [comment]
@vyasr Thank you so much for working on this!

File: .github/workflows/publish-packages.yml
Comment (vyasr) [reply .github/workflows/publish-packages.yml:30]: Sounds good, figured I'd put in the comments and see what reviewers thought.

File: .github/workflows/publish-packages.yml
Comment (vyasr) [reply .github/workflows/publish-packages.yml:22]: Sure.

File: .github/workflows/publish-packages.yml
Comment (vyasr) [reply .github/workflows/publish-packages.yml:24]: I had a comment earlier asking this question then removed it. Sure, I can set this to 3.10.

File: .github/workflows/deploy-pypi.yml
Comment (vyasr) [reply .github/workflows/deploy-pypi.yml:6]: I don't think you can really express that with a glob. The closest that you can get is maybe `v*.*.*`. I'll change to that, it is at least slightly more discriminating and doesn't automatically match any tag starting with `v`. That said, we don't really tag for anything other than releases so I'm not overly concerned (although I get the desire for a more precise match here).

File: .github/workflows/deploy-test-pypi.yml
Comment (vyasr) [reply .github/workflows/deploy-test-pypi.yml:6]: I did `release/*.*.*` for consistency with the tag filter.

File: .github/workflows/publish-packages.yml
Comment (vyasr) [reply .github/workflows/publish-packages.yml:38]: Oversight on my part, was trying to fix an issue at one point and this is a holdover.

**Review by csadorf — COMMENTED** [review]
Thanks for taking this on! Two suggestions.

**Comment by b-butler** [comment]
@bdice, @csadorf, @vyasr I went ahead and applied all suggested changes. I do not know how to test this without merging and updating `release/1.8.0`, but let me know if you have any more changes.

File: .github/workflows/publish-packages.yml
Comment (bdice) [reply .github/workflows/publish-packages.yml]: @b-butler In 51de0e252827553647d32e72e8c37ec730f30218 we lost the testing step. We need to test the actual wheel that we plan to upload. Relying on other tests (and not testing the final package) has resulted in issues in the past. Examples are when tests ran fine from the repo directory but submission scripts (signac-flow) or CSS files (signac-dashboard) were not properly bundled, meaning that the packaged wheel was incomplete.

**Review by bdice — APPROVED** [review]
I'm going to remove trigger-ability (see comments). I'd like a second approval from another maintainer, then we can merge.

File: .github/workflows/deploy-pypi.yml
Comment (bdice) [inline .github/workflows/deploy-pypi.yml:14]: I think we want to use the action in the current repo state, rather than a "versioned" copy. This currently shows an error:

image
https://github.com/glotzerlab/signac/actions/runs/3187228681/workflow

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_iduses
```suggestion
    uses: ./.github/workflows/publish-packages.yml
```

File: .github/workflows/deploy-test-pypi.yml
Comment (bdice) [inline .github/workflows/deploy-test-pypi.yml:14]:
```suggestion
    uses: ./.github/workflows/publish-packages.yml
```

File: .github/workflows/deploy-pypi.yml
Comment (bdice) [reply .github/workflows/deploy-pypi.yml]: Yup, this was broken by 14ad15f4945f2a49cbd748ae3c87153009db00df.

**Pull request merged**
